theano/compile/tests/test_function_module.py

```diff
@@ -239,6 +239,41 @@ def test_copy(self):
         f(1, 2)  # put them out of sync
         self.assertFalse(f(1, 2) == g(1, 2))  # they should not be equal anymore.
 
+    def test_copy_share_memory(self):
+        x = T.fscalar('x')
+        y = T.tanh((x+2)/(x-0.2)**2)
+
+        # test for PerformaLinker, will cover VM_linker later
+        ori = theano.function([x], [y], mode="FAST_COMPILE")
```

Comment:
I added test for ```FAST_RUN``` and ```FAST_COMPILE``` modes, which corresponds to ```PerformLinker``` and ```VM_Linker```. ```ProfileMode``` is deprecated so I didn't test it.

-----------------------------

For ```DebugMode``` which use ```_Linker```, the test didn't passed. The reason is that ```variable```s in ```storage_map``` generated by ```_Linker``` are copied variables in ```fgraph```. Therefore I cannot map old variable to new one using memo.
#### Test and output
I set a break_point inside copy() and type the following code.
```python
(Pdb) storage_map=self.fn.storage_map
(Pdb) [key in memo for key in storage_map]
[False, True, False, False, False, False, False, False, False, False, False, False, False, False, True, True, False, False, False, False, False, False, False, False, False, True, False, True, False, True, False, False, False, True, False, False, False, False, True, False]
```

#### Reason
```python
# LINE 1689 in theano/compile/debugmode.py
fgraph_equiv = fgraph.equivalence_tracker
order_outputs = copy.copy(fgraph_equiv.all_variables_ever) 
# ^^^^     variables are copied in the upper line. ^^^^
del fgraph_equiv
order_outputs.reverse()
order = graph.io_toposort(fgraph.inputs, order_outputs)
# variables in order are used to init storage_map
```
So I drop the test for DebugMode.
To solve this problem, I can either delete the copy line or add a doc to inform users not to use copy() in DebugMode.
I only had a glance to debugmode.py, I hope I didn't make it wrong.